File: isegoria/src/main/java/com/eulersbridge/isegoria/friends/FriendsViewModel.java
package com.eulersbridge.isegoria.friends;

import android.app.Application;
import android.arch.lifecycle.AndroidViewModel;
import android.arch.lifecycle.LiveData;
import android.arch.lifecycle.MutableLiveData;
import android.arch.lifecycle.Transformations;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.text.TextUtils;

import com.eulersbridge.isegoria.IsegoriaApp;
import com.eulersbridge.isegoria.network.api.models.Contact;
import com.eulersbridge.isegoria.network.api.models.FriendRequest;
import com.eulersbridge.isegoria.network.api.models.Institution;
import com.eulersbridge.isegoria.network.api.models.User;
import com.eulersbridge.isegoria.util.Strings;
import com.eulersbridge.isegoria.util.data.RetrofitLiveData;
import com.eulersbridge.isegoria.util.data.SingleLiveData;

import java.util.ArrayList;
import java.util.Iterator;
import java.util.List;

@SuppressWarnings("WeakerAccess")
public class FriendsViewModel extends AndroidViewModel {

    private LiveData<List<User>> searchResults;
    private LiveData<List<FriendRequest>> sentFriendRequests;
    private LiveData<List<FriendRequest>> receivedFriendRequests;
    private LiveData<List<Contact>> friends;

    final MutableLiveData<Boolean> searchSectionVisible = new MutableLiveData<>();
    final MutableLiveData<Boolean> sentRequestsVisible = new MutableLiveData<>();
    final MutableLiveData<Boolean> receivedRequestsVisible = new MutableLiveData<>();
    final MutableLiveData<Boolean> friendsVisible = new MutableLiveData<>();

    public FriendsViewModel(@NonNull Application application) {
        super(application);

        searchSectionVisible.setValue(false);
        sentRequestsVisible.setValue(false);
        receivedRequestsVisible.setValue(false);
        friendsVisible.setValue(true);
    }

    void onExit() {
        IsegoriaApp app = getApplication();
        app.friendsVisible.setValue(false);
    }

    /**
     * Convenience method to cancel a LiveData object if it exists and is a Retrofit API request.
     */
    private void cancelIfPossible(@Nullable LiveData liveData) {
        if (liveData != null && liveData instanceof RetrofitLiveData)
            ((RetrofitLiveData) liveData).cancel();
    }

    @Override
    protected void onCleared() {
        cancelIfPossible(friends);
        cancelIfPossible(sentFriendRequests);
        cancelIfPossible(receivedFriendRequests);
        cancelIfPossible(searchResults);
    }

    private void showSearch() {
        searchSectionVisible.setValue(true);
        sentRequestsVisible.setValue(false);
        receivedRequestsVisible.setValue(false);
        friendsVisible.setValue(false);
    }

    void hideSearch() {
        searchSectionVisible.setValue(false);

        boolean haveSentRequests = sentFriendRequests.getValue() != null
                && sentFriendRequests.getValue().size() > 0;

        sentRequestsVisible.setValue(haveSentRequests);

        boolean haveReceivedRequests = receivedFriendRequests.getValue() != null
                && receivedFriendRequests.getValue().size() > 0;

        receivedRequestsVisible.setValue(haveReceivedRequests);

        friendsVisible.setValue(true);
    }

    LiveData<List<User>> onSearchQueryChanged(String query) {
        showSearch();

        if (!TextUtils.isEmpty(query) && query.length() > 2) {
            IsegoriaApp app = getApplication();

            searchResults = new RetrofitLiveData<>(app.getAPI().searchForUsers(query));

        } else {
            searchResults = new SingleLiveData<>(null);
        }

        return searchResults;
    }

    LiveData<List<Contact>> getFriends() {
        if (friends == null) {
            IsegoriaApp app = getApplication();
            friends = new RetrofitLiveData<>(app.getAPI().getFriends());
        }

        return friends;
    }

    LiveData<List<FriendRequest>> getSentFriendRequests() {
        if (sentFriendRequests == null) {
            IsegoriaApp app = getApplication();

            return Transformations.switchMap(app.loggedInUser, user -> {
                if (user != null) {

                    LiveData<List<FriendRequest>> requests = new RetrofitLiveData<>(app.getAPI().getFriendRequestsSent(user.getId()));
                    sentFriendRequests = Transformations.switchMap(requests, sentFriendRequests -> {
                        sentRequestsVisible.setValue(sentFriendRequests != null && sentFriendRequests.size() > 0);

                        return new SingleLiveData<>(sentFriendRequests);
                    });

                    return sentFriendRequests;
                }

                return new SingleLiveData<>(null);
            });
        }

        return sentFriendRequests;
    }

    LiveData<List<FriendRequest>> getReceivedFriendRequests() {
        if (receivedFriendRequests == null) {
            IsegoriaApp app = getApplication();

            return Transformations.switchMap(app.loggedInUser, user -> {
                if (user != null) {

                    LiveData<List<FriendRequest>> requests = new RetrofitLiveData<>(app.getAPI().getFriendRequestsReceived(user.getId()));
                    receivedFriendRequests = Transformations.switchMap(requests, receivedFriendRequests -> {

                        List<FriendRequest> filteredReceivedRequests;
                        if (receivedFriendRequests != null) {
                            filteredReceivedRequests = new ArrayList<>(receivedFriendRequests);

                            Iterator<FriendRequest> iterator = filteredReceivedRequests.iterator();
                            while (iterator.hasNext()) {
                                FriendRequest request = iterator.next();

                                if (request.accepted != null)
                                    iterator.remove();
                            }

                        } else {
                            filteredReceivedRequests = new ArrayList<>();
                        }

                        receivedRequestsVisible.setValue(filteredReceivedRequests.size() > 0);

                        return new SingleLiveData<>(filteredReceivedRequests);
                    });

                    return receivedFriendRequests;
                }

                return new SingleLiveData<>(null);
            });
        }

        return receivedFriendRequests;
    }

    LiveData<Boolean> addFriend(@NonNull String newFriendEmail) {
        if (Strings.isValidEmail(newFriendEmail)) {

            IsegoriaApp app = getApplication();

            return Transformations.switchMap(app.loggedInUser, user -> {
                LiveData<Void> friendRequest = new RetrofitLiveData<>(app.getAPI().addFriend(user.email, newFriendEmail));
                return Transformations.switchMap(friendRequest, __ -> new SingleLiveData<>(true));
            });
        }

        return new SingleLiveData<>(false);
    }

    LiveData<Boolean> acceptFriendRequest(long requestId) {
        IsegoriaApp app = getApplication();

        LiveData<Void> friendRequest = new RetrofitLiveData<>(app.getAPI().acceptFriendRequest(requestId));
        return Transformations.switchMap(friendRequest, __ -> {
            getReceivedFriendRequests();
            getFriends();

            return new SingleLiveData<>(true);
        });
    }

    LiveData<Boolean> rejectFriendRequest(long requestId) {
        IsegoriaApp app = getApplication();

        LiveData<Void> friendRequest = new RetrofitLiveData<>(app.getAPI().rejectFriendRequest(requestId));
        return Transformations.switchMap(friendRequest, __ -> {
            getReceivedFriendRequests();
            getFriends();

            return new SingleLiveData<>(true);
        });
    }

    LiveData<Institution> getInstitution(long institutionId) {
        IsegoriaApp app = getApplication();

        return new RetrofitLiveData<>(app.getAPI().getInstitution(institutionId));
    }

}

File: isegoria/src/main/java/com/eulersbridge/isegoria/friends/FriendsViewModel.kt
package com.eulersbridge.isegoria.friends

import android.arch.lifecycle.LiveData
import android.arch.lifecycle.MutableLiveData
import android.arch.lifecycle.ViewModel
import com.eulersbridge.isegoria.AppRouter
import com.eulersbridge.isegoria.Repository
import com.eulersbridge.isegoria.network.api.model.Contact
import com.eulersbridge.isegoria.network.api.model.FriendRequest
import com.eulersbridge.isegoria.network.api.model.Institution
import com.eulersbridge.isegoria.network.api.model.User
import com.eulersbridge.isegoria.util.data.SingleLiveData
import com.eulersbridge.isegoria.util.extension.subscribeSuccess
import com.eulersbridge.isegoria.util.extension.toBooleanSingle
import com.eulersbridge.isegoria.util.extension.toLiveData
import io.reactivex.disposables.CompositeDisposable
import io.reactivex.rxkotlin.addTo
import javax.inject.Inject

class FriendsViewModel
@Inject constructor(
    private val appRouter: AppRouter,
    private val repository: Repository
) : ViewModel() {

    private val compositeDisposable = CompositeDisposable()

    internal var searchResults = MutableLiveData<List<User>>()
    internal var sentFriendRequests = MutableLiveData<List<FriendRequest>>()
    internal var receivedFriendRequests = MutableLiveData<List<FriendRequest>>()
    internal var friends = MutableLiveData<List<Contact>>()

    internal val searchSectionVisible = MutableLiveData<Boolean>()
    internal val sentRequestsVisible = MutableLiveData<Boolean>()
    internal val receivedRequestsVisible = MutableLiveData<Boolean>()
    internal val friendsVisible = MutableLiveData<Boolean>()

    init {
        searchSectionVisible.value = false
        sentRequestsVisible.value = false
        receivedRequestsVisible.value = false
        friendsVisible.value = true

        getFriends()
        getReceivedFriendRequests()
        getSentFriendRequests()
    }

    override fun onCleared() {
        compositeDisposable.dispose()
    }

    internal fun onDestroy() {
        appRouter.setFriendsScreenVisible(false)
    }

    private fun showSearch() {
        searchSectionVisible.value = true
        sentRequestsVisible.value = false
        receivedRequestsVisible.value = false
        friendsVisible.value = false
    }

    internal fun hideSearch() {
        searchSectionVisible.value = false

        val haveSentRequests = sentFriendRequests.value?.isNotEmpty() ?: false
        sentRequestsVisible.value = haveSentRequests

        val haveReceivedRequests = receivedFriendRequests.value?.isNotEmpty() ?: false
        receivedRequestsVisible.value = haveReceivedRequests

        friendsVisible.value = true
    }

    internal fun onSearchQueryChanged(query: String) {
        showSearch()

        if (!query.isBlank() && query.length > 2) {
            repository.searchForUsers(query).subscribeSuccess {
                searchResults.postValue(it)
            }

        } else {
            searchResults.value = emptyList()
        }
    }

    private fun getFriends() {
        repository.getFriends().subscribeSuccess {
            friends.postValue(it)
        }.addTo(compositeDisposable)
    }

    private fun getSentFriendRequests() {
        repository.getSentFriendRequests().subscribeSuccess {
            sentFriendRequests.postValue(it)
            sentRequestsVisible.postValue(it.isNotEmpty())
        }.addTo(compositeDisposable)
    }

    private fun getReceivedFriendRequests() {
        repository.getReceivedFriendRequests().subscribeSuccess {
            receivedFriendRequests.postValue(it)
            receivedRequestsVisible.postValue(it.isNotEmpty())
        }.addTo(compositeDisposable)
    }

    internal fun addFriend(newFriendEmail: String): LiveData<Boolean> {
        if (!newFriendEmail.isBlank())
            repository.addFriend(newFriendEmail).toLiveData()

        return SingleLiveData(false)
    }

    internal fun acceptFriendRequest(requestId: Long): LiveData<Boolean> {
        return repository.acceptFriendRequest(requestId).doOnComplete {
            getReceivedFriendRequests()
            getFriends()
        }.toBooleanSingle().toLiveData()
    }

    internal fun rejectFriendRequest(requestId: Long): LiveData<Boolean> {
        return repository.rejectFriendRequest(requestId).doOnComplete {
            getReceivedFriendRequests()
            getFriends()
        }.toBooleanSingle().toLiveData()
    }

    //TODO: Fix
    internal fun getInstitution(institutionId: Long): LiveData<Institution>
            = repository.getInstitution(institutionId).toLiveData()

}
